{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1a1495df",
   "metadata": {},
   "source": [
    "## Data Source and Reference Set Construction  \n",
    "Anchor sequences were obtained from the Spider Silkome Database (https://spider-silkome.org/about).  \n",
    "\n",
    "### Environment Setup  \n",
    "Ensure that the following dependencies have been installed (via Pixi):\n",
    "```bash\n",
    "pixi init spider_silkome\n",
    "pixi add python pandas biopython mafft hmmer samtools flye minimap2 seqkit cd-hit\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "594cae69",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import glob\n",
    "import pandas as pd\n",
    "from spider_silkome_module.config import PROCESSED_DATA_DIR, RAW_DATA_DIR, SCRIPTS_DIR, EXTERNAL_DATA_DIR, INTERIM_DATA_DIR\n",
    "\n",
    "from spider_silkome_module import run_shell_command_with_check\n",
    "from typing import List, Dict, Optional, Tuple\n",
    "from dataclasses import dataclass\n",
    "\n",
    "# Global Configuration\n",
    "class Config:\n",
    "    \"\"\"\n",
    "    Configuration class to hold paths and thresholds.\n",
    "    Using a centralized config makes modifications easier.\n",
    "    \"\"\"\n",
    "    PROJECT_NAME = \"spider_silkome_20251222\"\n",
    "    # Paths\n",
    "    SEEDS_QC_DIR = INTERIM_DATA_DIR / PROJECT_NAME / \"spidroin_seeds_qc\"\n",
    "    HMMBUILD_INPUT_DIR = INTERIM_DATA_DIR / PROJECT_NAME / \"hmmbuild_input\"\n",
    "    HMMBUILD_OUTPUT_DIR = INTERIM_DATA_DIR / PROJECT_NAME / \"hmmbuild_output\"\n",
    "    HMMSEARCH_OUTPUT_DIR = INTERIM_DATA_DIR / PROJECT_NAME / \"hmmserch_output\"\n",
    "    MINIPROT_OUTPUT_DIR = INTERIM_DATA_DIR / PROJECT_NAME / \"miniprot_output\"\n",
    "    spider_genome_path = RAW_DATA_DIR / \"spider_genome\"\n",
    "    spider_genomes = [\"Araneus_ventricosus\",\n",
    "                    \"Pardosa_pseudoannulata\",\n",
    "                    \"Evarcha_sp\",\n",
    "                    \"Pholcus_sp\",\n",
    "                    \"Heteropoda_venatoria\",\n",
    "                    \"Scorpiops_zhui\",\n",
    "                    \"Hippasa_lycosina\",\n",
    "                    \"Songthela_sp\",\n",
    "                    \"Pandercetes_sp\",\n",
    "                    \"Trichonephila_clavata\"\n",
    "    ]\n",
    "\n",
    "\n",
    "\n",
    "    # Threads\n",
    "    THREADS = 8\n",
    "\n",
    "    # Thresholds\n",
    "    E_VALUE_THRES = 1e-10\n",
    "    MIN_GENE_LEN = 500       # Minimum distance between NTD and CTD\n",
    "    MAX_GENE_LEN = 100000    # Maximum distance (100kb)\n",
    "\n",
    "    def __init__(self):\n",
    "        # Create directories if they don't exist\n",
    "        os.makedirs(Config.SEEDS_QC_DIR, exist_ok=True)\n",
    "        os.makedirs(Config.HMMBUILD_INPUT_DIR, exist_ok=True)\n",
    "        os.makedirs(Config.HMMBUILD_OUTPUT_DIR, exist_ok=True)\n",
    "        os.makedirs(Config.HMMSEARCH_OUTPUT_DIR, exist_ok=True)\n",
    "        os.makedirs(Config.MINIPROT_OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Initialize configuration\n",
    "config = Config()\n",
    "\n",
    "spidroin_seeds = EXTERNAL_DATA_DIR / \"spidroin_seeds_collection.faa\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55704049",
   "metadata": {},
   "source": [
    "## Acquisition and cleaning of seed sequences\n",
    "\n",
    "The spidroin family is large and encompasses multiple types, including major ampullate silk (MaSp), minor ampullate silk (MiSp), flagelliform silk (Flag), aciniform silk (AcSp), pyriform silk (PySp), tubuliform silk (TuSp/CySp), and aggregate silk (AgSp). In addition, Schöneberg et al. (2025) reported a unique spidroin type found in the Araneoidea.\n",
    "\n",
    "Procedure:\n",
    "\n",
    "1. Download and classification:  \n",
    "   Download the NTD and CTD amino acid sequences of each subfamily from the database.\n",
    "\n",
    "2. Quality control:  \n",
    "   Remove sequences containing simple repeats or with abnormal lengths (<50 aa).\n",
    "\n",
    "3. Redundancy removal:  \n",
    "   Use the CD-HIT tool at a 95% similarity threshold to eliminate highly redundant sequences and prevent model overfitting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e96455e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quality Control\n",
    "qc_cmd = f\"pixi run python3 {SCRIPTS_DIR}/extract_terminal_domains.py {spidroin_seeds} -o {config.SEEDS_QC_DIR} -l 50 --similarity 0.5\"\n",
    "run_shell_command_with_check(qc_cmd, config.SEEDS_QC_DIR/\"processing_report.tsv\", force=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f1c8e4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove highly redundant sequences using cd-hit\n",
    "for input_path in Config.SEEDS_QC_DIR.glob(\"*TD.faa\"):\n",
    "    output_path = input_path.with_stem(f\"{input_path.stem}_de_dup\")\n",
    "    de_dup_cmd = f\"pixi run cd-hit -i {input_path} -o {output_path} -c 0.95 -T 0 -M 0\"\n",
    "    run_shell_command_with_check(de_dup_cmd, output_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54228d5b",
   "metadata": {},
   "source": [
    "## Multiple sequence alignment and pHMM construction  \n",
    "\n",
    "To improve search sensitivity, especially for distantly related species (such as the span from Nephila clavata to spiders of the suborder Araneomorphae), it is necessary to construct a profile HMM.  \n",
    "\n",
    "- Technical details:  \n",
    "  - Tool selection: HMMER 3.4 suite (hmmbuild).  \n",
    "  - Alignment strategy: Perform multiple sequence alignment (MSA) using the L-INS-i algorithm in MAFFT. This algorithm is the most accurate for capturing local structural features (such as conserved cysteine residues in the NTD or specific α-helical structures).  \n",
    "- Model training:  \n",
    "  - Construction of subfamily-specific models (e.g., MaSp_NTD.hmm): used for precise classification.  \n",
    "  - Calibration: Use hmmpress to perform binary compression and indexing of the models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "655f1694",
   "metadata": {},
   "outputs": [],
   "source": [
    "# multiple sequence alignment\n",
    "for input_path in Config.SEEDS_QC_DIR.glob(\"*_de_dup.faa\"):\n",
    "    name = input_path.stem.split(\"_de\")[0]\n",
    "    mafft_output_path = Config.HMMBUILD_INPUT_DIR / f\"{name}_aln.faa\"\n",
    "    mafft_cmd = f\"mafft --maxiterate 1000 --localpair --thread -1 {input_path} > {mafft_output_path}\"\n",
    "    run_shell_command_with_check(mafft_cmd, mafft_output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "600efa94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alignment trimming\n",
    "for input_path in Config.HMMBUILD_INPUT_DIR.glob(\"*_aln.faa\"):\n",
    "    aln_trimed_output_path = input_path.with_name(input_path.stem + \"_trimed.faa\")\n",
    "    aln_trimed_cmd = f\"pixi run trimal -in {input_path} -out {aln_trimed_output_path} -gt 0.8 -cons 60\"\n",
    "    run_shell_command_with_check(aln_trimed_cmd, aln_trimed_output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a43daf0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# HMMER search\n",
    "for input_path in Config.HMMBUILD_INPUT_DIR.glob(\"*_aln_trimed.faa\"):\n",
    "    name = input_path.stem.split(\"_aln\")[0]\n",
    "    hmmbuild_output_path = Config.HMMBUILD_OUTPUT_DIR / f\"{name}.hmm\"\n",
    "    hmmbuild_cmd = f\"hmmbuild -n {name} --amino --cpu 70 {hmmbuild_output_path} {input_path}\"\n",
    "    run_shell_command_with_check(hmmbuild_cmd, hmmbuild_output_path)\n",
    "    hmmpress_cmd = f\"hmmpress {hmmbuild_output_path}\"\n",
    "    run_shell_command_with_check(hmmpress_cmd, f\"{hmmbuild_output_path}.h3m\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10db2fc8",
   "metadata": {},
   "source": [
    "## Genomic Scanning Strategy\n",
    "\n",
    "Conventional BLAST searches often produce a large number of false negatives when dealing with highly divergent sequences. In contrast, nhmmer (a component of HMMER) can directly map a protein pHMM model onto nucleotide sequences, or perform DNA-to-DNA HMM searches, thereby greatly improving sensitivity while maintaining high specificity.\n",
    "\n",
    "Implementation scheme:\n",
    "- Whole-genome scanning: Use nhmmer to perform a six-frame translation search against the target genome (or a direct nucleotide search, depending on the model type).\n",
    "- Parameter settings: Set the E-value threshold to \\(1 \\times 10^{-10}\\) to ensure high confidence; require --qcov (query coverage) greater than 70% to ensure that complete domains, rather than fragments, are detected.\n",
    "- Result filtering: Remove low-complexity, nonspecific matches. Spider silk protein genomes often contain large numbers of simple repeat sequences, which necessitates soft-filtering of HMM hits using the output of RepeatMasker."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac3122af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Translate protein based on GFF file\n",
    "for input_path in Config.HMMBUILD_OUTPUT_DIR.glob(\"*.hmm\"):\n",
    "    output_path = input_path.with_name(input_path.stem + \"_consensus.fa\")\n",
    "    cmd = f\"pixi run hmmemit -c {input_path} > {output_path}\"\n",
    "    run_shell_command_with_check(cmd, output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6556c755",
   "metadata": {},
   "outputs": [],
   "source": [
    "for spider in Config.spider_genomes:\n",
    "    hmm_dir = Config.HMMBUILD_OUTPUT_DIR\n",
    "    spider_genome = Config.spider_genome_path / spider / f\"{spider}.fa\"\n",
    "    spider_protein = spider_genome.with_name(spider + \".protein.fasta\")\n",
    "    spider_gff = spider_genome.with_name(spider + \".gff\")\n",
    "    gffread_cmd = f\"pixi run gffread -g {spider_genome} -y {spider_protein} {spider_gff}\"\n",
    "    run_shell_command_with_check(gffread_cmd, spider_protein)\n",
    "    spider_hmmsearch_output_dir = Config.HMMSEARCH_OUTPUT_DIR / spider\n",
    "    for hmm in hmm_dir.glob(\"*.hmm\"):\n",
    "        hmmsearch_cmd = f\"pixi run hmmsearch --cpu 75 --tblout {spider_hmmsearch_output_dir / hmm.stem}.tbl --domtblout {spider_hmmsearch_output_dir / hmm.stem}.domtbl {hmm} {spider_protein} > {spider_hmmsearch_output_dir / hmm.stem}.out\"\n",
    "        run_shell_command_with_check(hmmsearch_cmd, spider_hmmsearch_output_dir / f\"{hmm.stem}.out\", force=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
