{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "77bf3519",
   "metadata": {},
   "source": [
    "# Automated Spidroin Annotation\n",
    "\n",
    "This project base on the github repository of [Spidroins](https://github.com/amandamarkee/spidroins). The goal of this project is to develop an automated pipeline for spidroin annotation for spider genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8f40eecd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2025-11-28 14:10:10.927\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.config\u001b[0m:\u001b[36m<module>\u001b[0m:\u001b[36m11\u001b[0m - \u001b[1mPROJ_ROOT path is: /home/gyk/project/spider_silkome\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "from spider_silkome_module import (\n",
    "    RAW_DATA_DIR,\n",
    "    INTERIM_DATA_DIR,\n",
    "    PROCESSED_DATA_DIR,\n",
    "    EXTERNAL_DATA_DIR,\n",
    "    SCRIPTS_DIR\n",
    ")\n",
    "from spider_silkome_module import (\n",
    "    run_shell_command_with_check,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0e92141",
   "metadata": {},
   "source": [
    "## Configure the environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eac9ad91",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_name = \"automated_spidroin_annotation\"\n",
    "spider = \"Trichonephila_clavata\"\n",
    "\n",
    "# Step 1: Identifying and sorting spidroins\n",
    "os.makedirs(INTERIM_DATA_DIR / project_name / spider, exist_ok=True)\n",
    "spider_genome_file = RAW_DATA_DIR / f\"spider_genome/{spider}.fa.gz\"\n",
    "spidroin_fasta_file = EXTERNAL_DATA_DIR / \"spider-silkome-database.v1.prot.fixed.fasta\"\n",
    "gnome_mpi = INTERIM_DATA_DIR / project_name / spider / f\"{spider}.mpi\"\n",
    "miniprot_out = INTERIM_DATA_DIR / project_name / spider / \"miniprot_out.gff\"\n",
    "miniprot_out_mRNA = miniprot_out.with_suffix(\".mRNA.gff\")\n",
    "miniprot_out_mRNA_expanded = miniprot_out_mRNA.with_suffix(\".expanded.csv\")\n",
    "\n",
    "miniprot_to_coordinates_script = SCRIPTS_DIR / \"miniprot_to_coordinates.py\"\n",
    "miniprot_out_dedup = INTERIM_DATA_DIR / project_name / spider / \"miniprot_out_dedup.tsv\"\n",
    "miniprot_out_paired = INTERIM_DATA_DIR / project_name / spider / \"miniprot_out_paired.bed\"\n",
    "miniprot_out_unpaired = INTERIM_DATA_DIR / project_name / spider / \"miniprot_out_unpaired.bed\"\n",
    "bin_size = 500\n",
    "pad = 1000\n",
    "min_distance = 15000\n",
    "max_distance = 90000\n",
    "min_positive = 0.75\n",
    "ref_fasta = spider_genome_file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41361d75",
   "metadata": {},
   "source": [
    "## Step 1: Identifying and Sorting Spidroins\n",
    "\n",
    "In this step, we will use the blastn to map the sppider genome to the spidroin N/C terminal protein sequence that can produce a file including qseqid, sseqid, pident, length, mismatch, gapopen, qstart, qend, sstart, send, evalue, bitscore."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "948e4263",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2025-11-28 14:11:56.029\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.features\u001b[0m:\u001b[36mrun_shell_command_with_check\u001b[0m:\u001b[36m46\u001b[0m - \u001b[1mThe output file already exists, skipping execution: /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/Trichonephila_clavata.mpi\u001b[0m\n",
      "\u001b[32m2025-11-28 14:11:56.030\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.features\u001b[0m:\u001b[36mrun_shell_command_with_check\u001b[0m:\u001b[36m46\u001b[0m - \u001b[1mThe output file already exists, skipping execution: /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out.gff\u001b[0m\n",
      "\u001b[32m2025-11-28 14:11:56.030\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.features\u001b[0m:\u001b[36mrun_shell_command_with_check\u001b[0m:\u001b[36m46\u001b[0m - \u001b[1mThe output file already exists, skipping execution: /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out.mRNA.gff\u001b[0m\n",
      "\u001b[32m2025-11-28 14:11:56.031\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.features\u001b[0m:\u001b[36mrun_shell_command_with_check\u001b[0m:\u001b[36m50\u001b[0m - \u001b[1mExecute command: python /home/gyk/project/spider_silkome/scripts/miniprot_to_coordinates.py --miniprot-gff /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out.mRNA.gff --out-dedup /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out_dedup.tsv --out-paired /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out_paired.bed --out-unpaired /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out_unpaired.bed --bin-size 500 --pad 1000 --min-distance 15000 --max-distance 90000 --ref-fasta /home/gyk/project/spider_silkome/data/raw/spider_genome/Trichonephila_clavata.fa.gz --min-positive 0.75\u001b[0m\n",
      "Reading miniprot GFF: /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out.mRNA.gff\n",
      "  Found 12071 mRNA records\n",
      "  1483 hits after positive >= 0.75 filter\n",
      "Deduplicating hits (bin size: 500 bp)...\n",
      "  237 hits after deduplication\n",
      "Pairing N/C terminals (distance: 15000-90000 bp)...\n",
      "  11 paired genes\n",
      "  215 unpaired hits\n",
      "Wrote deduplicated hits to: /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out_dedup.tsv\n",
      "Wrote paired genes (with 1000bp padding) to: /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out_paired.bed\n",
      "Wrote unpaired hits (no padding) to: /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out_unpaired.bed\n",
      "\u001b[32m2025-11-28 14:12:41.136\u001b[0m | \u001b[32m\u001b[1mSUCCESS \u001b[0m | \u001b[36mspider_silkome_module.features\u001b[0m:\u001b[36mrun_shell_command_with_check\u001b[0m:\u001b[36m53\u001b[0m - \u001b[32m\u001b[1mCommand executed successfully, output file: /home/gyk/project/spider_silkome/data/interim/automated_spidroin_annotation/Trichonephila_clavata/miniprot_out_dedup.tsv\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "index_cmd = f\"miniprot -t70 -d {gnome_mpi} {spider_genome_file}\"\n",
    "run_shell_command_with_check(index_cmd, gnome_mpi)\n",
    "\n",
    "align_cmd = f\"miniprot -t 70 -I --gff {gnome_mpi} {spidroin_fasta_file} > {miniprot_out}\"\n",
    "run_shell_command_with_check(align_cmd, miniprot_out)\n",
    "\n",
    "grep_cmd = f\"grep 'mRNA' {miniprot_out} | sort -k1,1V -k4,4n > {miniprot_out_mRNA}\"\n",
    "run_shell_command_with_check(grep_cmd, miniprot_out_mRNA)\n",
    "\n",
    "miniprot_to_coordinates_cmd = f\"python {miniprot_to_coordinates_script} --miniprot-gff {miniprot_out_mRNA} --out-dedup {miniprot_out_dedup} --out-paired {miniprot_out_paired} --out-unpaired {miniprot_out_unpaired} --bin-size {bin_size} --pad {pad} --min-distance {min_distance} --max-distance {max_distance} --ref-fasta {ref_fasta} --min-positive {min_positive}\"\n",
    "run_shell_command_with_check(miniprot_to_coordinates_cmd, miniprot_out_dedup, force=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a12373ab",
   "metadata": {},
   "source": [
    "Organize and sort the output files of miniprot to facilitate the next step of processing.\n",
    "\n",
    "<zh>对 miniprot 的输出文件进行整理和排序，方便下一步的整理</zh>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56cec651",
   "metadata": {},
   "outputs": [],
   "source": [
    "miniprot_out_mRNA_table = pd.read_csv(miniprot_out_mRNA, sep=\"\\t\")\n",
    "gff_header = [\"seqid\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attributes\"]\n",
    "miniprot_out_mRNA_table.columns = gff_header\n",
    "\n",
    "def parse_attributes(attr):\n",
    "    items = [x.split(\"=\") for x in attr.split(\";\") if \"=\" in x]\n",
    "    return {k: v for k, v in items}\n",
    "\n",
    "df_attrs = miniprot_out_mRNA_table[\"attributes\"].apply(parse_attributes).apply(pd.Series)\n",
    "miniprot_out_mRNA_table_expanded = miniprot_out_mRNA_table.drop(columns=[\"attributes\"]).join(df_attrs)\n",
    "\n",
    "miniprot_out_mRNA_table_expanded.to_csv(miniprot_out_mRNA_expanded, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
