{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4c6d2f56",
   "metadata": {},
   "source": [
    "# DRS Mapping\n",
    "Use IsoQuant software to align DRS data to the genome, calibrate it based on second-generation sequencing, and then obtain transcript boundary information.\n",
    "\n",
    "<zh>使用 IsoQuant 软件把 DRS 数据比对到基因组上，并且根据二代测序进行校准，然后获取转录本边界信息。</zh>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "af4c83ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2025-10-17 13:22:41.294\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.config\u001b[0m:\u001b[36m<module>\u001b[0m:\u001b[36m11\u001b[0m - \u001b[1mPROJ_ROOT path is: /home/gyk/project/spider_silkome\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "from spider_silkome_module import (\n",
    "    RAW_DATA_DIR,\n",
    "    INTERIM_DATA_DIR,\n",
    "    PROCESSED_DATA_DIR,\n",
    ")\n",
    "from spider_silkome_module import (\n",
    "    run_shell_command_with_check,\n",
    ")\n",
    "\n",
    "from spider_silkome_module import (\n",
    "    GeneralGFF,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5812362",
   "metadata": {},
   "source": [
    "## Create Genome STAR index\n",
    "\n",
    "Note: STAR needs to be installed in advance. Download link: https://github.com/alexdobin/STAR\n",
    "\n",
    "<zh>注意：需要提前安装好 STAR，下载地址：https://github.com/alexdobin/STAR</zh>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b161cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_file = f\"{RAW_DATA_DIR}/spider_genome/Trichonephila_clavata.fa\"\n",
    "spider = \"Trichonephila_clavata\"\n",
    "# genome_index_dir = f\"{INTERIM_DATA_DIR}/star_index/{spider}\"\n",
    "# os.makedirs(genome_index_dir, exist_ok=True)\n",
    "# star_index_cmd = f\"STAR --runThreadN 70 --runMode genomeGenerate --genomeDir {genome_index_dir} --genomeFastaFiles {genome_file}\"\n",
    "# subprocess.run(star_index_cmd, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84f90231",
   "metadata": {},
   "source": [
    "## Creat BGI RNA-seq Bam files\n",
    "\n",
    "**Note:** Maker sure the nextflow was installed in your system.\n",
    "\n",
    "In this section, we use nf-core/rnaseq to create Bam files for DRS data.\n",
    "\n",
    "Please prepare `nf-params.json` and `samplesheet.csv` file in `RNA-seq_workflow` directory according the [nf-core/rnaseq document](https://nf-co.re/rnaseq).\n",
    "\n",
    "Run `nextflow run nf-core/rnaseq -r 3.19.0 -name BGI_RNA-seq -profile docker -params-file nf-params.json` in `RNA-seq_workflow` directory."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c92e50e",
   "metadata": {},
   "source": [
    "## DRS Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71b92be9",
   "metadata": {},
   "source": [
    "Some errors occurred when running DRS mapping. So the gff file must be fixed by `agat_sp_fix_features_locations_duplicated.pl` before running the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b87d0769",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "!awk -F'\\t' '$3 ~ /^(gene|mRNA|exon|CDS)$/' /home/gyk/project/spider_silkome/data/raw/spider_genome/Trichonephila_clavata.gff > /home/gyk/project/spider_silkome/data/raw/spider_genome/Trichonephila_clavata_fixed.gff\n",
    "!pixi run --environment agat agat_convert_sp_gxf2gxf.pl \\\n",
    "    --gff /home/gyk/project/spider_silkome/data/raw/spider_genome/Trichonephila_clavata_fixed.gff \\\n",
    "    --output /home/gyk/project/spider_silkome/data/raw/spider_genome/Trichonephila_clavata_fixed_agat.gff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be902876",
   "metadata": {},
   "source": [
    "\n",
    "Note: if ValueError: Duplicate ID xxxxxx, then fixed the id manually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a780944e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2025-10-17 14:23:42.026\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.features\u001b[0m:\u001b[36mrun_shell_command_with_check\u001b[0m:\u001b[36m50\u001b[0m - \u001b[1mExecute command: isoquant.py --genedb /home/gyk/project/spider_silkome/data/raw/spider_genome/Trichonephila_clavata_fixed_agat.gff --reference /home/gyk/project/spider_silkome/data/raw/spider_genome/Trichonephila_clavata.fa --fastq /home/gyk/project/spider_silkome/data/raw/Tclav-F1_Ar-28-Trcl-f/pass.fq.gz --data_type nanopore -o /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant\u001b[0m\n",
      "2025-10-17 14:23:42,538 - INFO - Running IsoQuant version 3.9.0\n",
      "2025-10-17 14:23:42,538 - WARNING - Output folder already contains a previous run, some files may be overwritten. Use --resume to resume a failed run. Use --force to avoid this message.\n",
      "2025-10-17 14:23:42,538 - WARNING - Press Ctrl+C to interrupt the run now.\n",
      "2025-10-17 14:23:51,548 - INFO - Overwriting the previous run\n",
      "2025-10-17 14:23:52,550 - WARNING - /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT folder already exists, some files may be overwritten\n",
      "2025-10-17 14:23:52,550 - WARNING - /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/aux folder already exists, some files may be overwritten\n",
      "2025-10-17 14:23:52,553 - INFO - Novel unspliced transcripts will not be reported, set --report_novel_unspliced true to discover them\n",
      "2025-10-17 14:23:52,554 - INFO -  === IsoQuant pipeline started === \n",
      "2025-10-17 14:23:52,554 - INFO - Python version: 3.10.18 | packaged by conda-forge | (main, Jun  4 2025, 14:45:41) [GCC 13.3.0]\n",
      "2025-10-17 14:23:52,554 - INFO - gffutils version: 0.13\n",
      "2025-10-17 14:23:52,554 - INFO - pysam version: 0.23.0\n",
      "2025-10-17 14:23:52,554 - INFO - pyfaidx version: 0.8.1.3\n",
      "2025-10-17 14:23:52,554 - INFO - Reading reference genome from /home/gyk/project/spider_silkome/data/raw/spider_genome/Trichonephila_clavata.fa\n",
      "2025-10-17 14:23:52,555 - INFO - Checking input gene annotation\n",
      "2025-10-17 14:23:54,241 - INFO - Gene annotation seems to be correct\n",
      "2025-10-17 14:23:54,241 - INFO - Converting gene annotation file to .db format (takes a while)...\n",
      "2025-10-17 14:25:01,313 - INFO - Gene database written to /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/Trichonephila_clavata_fixed_agat.db\n",
      "2025-10-17 14:25:01,313 - INFO - Provide this database next time to avoid excessive conversion\n",
      "2025-10-17 14:25:01,314 - INFO - Indexing reference\n",
      "2025-10-17 14:26:31,110 - INFO - Converting gene annotation file /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/Trichonephila_clavata_fixed_agat.db to .bed format\n",
      "2025-10-17 14:27:01,614 - INFO - Gene database BED written to /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/Trichonephila_clavata_fixed_agat.bed\n",
      "2025-10-17 14:27:01,641 - INFO - Aligning /home/gyk/project/spider_silkome/data/raw/Tclav-F1_Ar-28-Trcl-f/pass.fq.gz to the reference, alignments will be saved to /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/aux/OUT_pass_8fbdf8_29e052_abff53.bam\n",
      "2025-10-17 14:27:01,643 - INFO - Running minimap2 version 2.30-r1287 (takes a while)\n",
      "2025-10-17 16:16:36,036 - INFO - Sorting alignments\n",
      "2025-10-17 16:18:07,962 - INFO - Indexing alignments\n",
      "2025-10-17 16:19:35,662 - INFO - Loading gene database from /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/Trichonephila_clavata_fixed_agat.db\n",
      "2025-10-17 16:19:35,663 - INFO - Loading reference genome from /home/gyk/project/spider_silkome/data/raw/spider_genome/Trichonephila_clavata.fa\n",
      "2025-10-17 16:20:03,824 - INFO - Processing 1 experiment\n",
      "2025-10-17 16:20:03,824 - INFO - Secondary alignments will not be used\n",
      "2025-10-17 16:20:03,824 - INFO - Processing experiment OUT\n",
      "2025-10-17 16:20:03,824 - INFO - Experiment has 1 BAM file: /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/aux/OUT_pass_8fbdf8_29e052_abff53.bam\n",
      "2025-10-17 16:20:17,829 - INFO - Collecting read alignments\n",
      "2025-10-17 16:20:17,926 - INFO - Processing chromosome Chr5\n",
      "2025-10-17 16:20:17,926 - INFO - Processing chromosome Chr8\n",
      "2025-10-17 16:20:17,926 - INFO - Processing chromosome Chr4\n",
      "2025-10-17 16:20:17,926 - INFO - Processing chromosome Chr1\n",
      "2025-10-17 16:20:17,926 - INFO - Processing chromosome Chr6\n",
      "2025-10-17 16:20:17,926 - INFO - Processing chromosome Chr9\n",
      "2025-10-17 16:20:17,926 - INFO - Processing chromosome Chr10\n",
      "2025-10-17 16:20:17,926 - INFO - Processing chromosome Chr11\n",
      "2025-10-17 16:20:17,927 - INFO - Processing chromosome Chr3\n",
      "2025-10-17 16:20:17,928 - INFO - Processing chromosome Chr7\n",
      "2025-10-17 16:20:17,928 - INFO - Processing chromosome Chr13\n",
      "2025-10-17 16:20:17,928 - INFO - Processing chromosome Chr12\n",
      "2025-10-17 16:20:17,929 - INFO - Processing chromosome Chr2\n",
      "2025-10-17 16:24:18,398 - INFO - Finished processing chromosome Chr12\n",
      "2025-10-17 16:25:30,934 - INFO - Finished processing chromosome Chr13\n",
      "2025-10-17 16:27:26,706 - INFO - Finished processing chromosome Chr9\n",
      "2025-10-17 16:28:16,544 - INFO - Finished processing chromosome Chr11\n",
      "2025-10-17 16:28:28,724 - INFO - Finished processing chromosome Chr8\n",
      "2025-10-17 16:28:35,001 - INFO - Finished processing chromosome Chr7\n",
      "2025-10-17 16:28:52,252 - INFO - Finished processing chromosome Chr10\n",
      "2025-10-17 16:29:06,975 - INFO - Finished processing chromosome Chr2\n",
      "2025-10-17 16:29:09,058 - INFO - Finished processing chromosome Chr1\n",
      "2025-10-17 16:29:32,493 - INFO - Finished processing chromosome Chr6\n",
      "2025-10-17 16:29:35,661 - INFO - Finished processing chromosome Chr5\n",
      "2025-10-17 16:30:05,577 - INFO - Finished processing chromosome Chr4\n",
      "2025-10-17 16:31:15,650 - INFO - Finished processing chromosome Chr3\n",
      "2025-10-17 16:31:15,712 - INFO - Counting multimapped reads\n",
      "2025-10-17 16:31:15,712 - INFO - Counting reads from Chr1\n",
      "2025-10-17 16:31:15,712 - INFO - Counting reads from Chr2\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr3\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr4\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr5\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr6\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr7\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr8\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr9\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr10\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr11\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr12\n",
      "2025-10-17 16:31:15,713 - INFO - Counting reads from Chr13\n",
      "2025-10-17 16:31:15,713 - INFO - Resolving multimappers\n",
      "2025-10-17 16:31:15,713 - INFO - Multimappers resolved\n",
      "2025-10-17 16:31:15,723 - INFO - Alignments collected, overall alignment statistics:\n",
      "2025-10-17 16:31:15,723 - INFO - primary: 5983631\n",
      "2025-10-17 16:31:15,723 - INFO - secondary: 1456915\n",
      "2025-10-17 16:31:15,724 - INFO - supplementary: 20478\n",
      "2025-10-17 16:31:15,724 - INFO - unaligned: 590092\n",
      "2025-10-17 16:31:15,724 - INFO - Finishing read assignment, total assignments 5805109, polyA percentage 27.2\n",
      "2025-10-17 16:31:15,725 - INFO - Read assignments files saved to /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/aux/OUT.save*. \n",
      "2025-10-17 16:31:15,726 - INFO - To keep these intermediate files for debug purposes use --keep_tmp flag\n",
      "2025-10-17 16:31:15,726 - INFO - Total assignments used for analysis: 5805109, polyA tail detected in 1578478 (27.2%)\n",
      "2025-10-17 16:31:15,726 - INFO - Processing assigned reads OUT\n",
      "2025-10-17 16:31:15,726 - INFO - Transcript models construction is turned on\n",
      "2025-10-17 16:31:15,727 - INFO - Transcript construction options:\n",
      "2025-10-17 16:31:15,727 - INFO -   Novel monoexonic transcripts will be reported: no\n",
      "2025-10-17 16:31:15,727 - INFO -   PolyA tails are required for multi-exon transcripts to be reported: no\n",
      "2025-10-17 16:31:15,727 - INFO -   PolyA tails are required for 2-exon transcripts to be reported: yes\n",
      "2025-10-17 16:31:15,727 - INFO -   PolyA tails are required for known monoexon transcripts to be reported: yes\n",
      "2025-10-17 16:31:15,727 - INFO -   PolyA tails are required for novel monoexon transcripts to be reported: yes\n",
      "2025-10-17 16:31:15,727 - INFO -   Splice site reporting level: only_canonical\n",
      "2025-10-17 16:31:15,802 - INFO - Processing chromosome Chr1\n",
      "2025-10-17 16:31:15,803 - INFO - Processing chromosome Chr2\n",
      "2025-10-17 16:31:15,803 - INFO - Processing chromosome Chr4\n",
      "2025-10-17 16:31:15,803 - INFO - Processing chromosome Chr3\n",
      "2025-10-17 16:31:15,803 - INFO - Processing chromosome Chr6\n",
      "2025-10-17 16:31:15,803 - INFO - Processing chromosome Chr5\n",
      "2025-10-17 16:31:15,803 - INFO - Processing chromosome Chr7\n",
      "2025-10-17 16:31:15,804 - INFO - Processing chromosome Chr8\n",
      "2025-10-17 16:31:15,804 - INFO - Processing chromosome Chr10\n",
      "2025-10-17 16:31:15,804 - INFO - Processing chromosome Chr11\n",
      "2025-10-17 16:31:15,804 - INFO - Processing chromosome Chr12\n",
      "2025-10-17 16:31:15,804 - INFO - Processing chromosome Chr13\n",
      "2025-10-17 16:31:15,804 - INFO - Processing chromosome Chr9\n",
      "2025-10-17 16:32:20,889 - INFO - Finished processing chromosome Chr13\n",
      "2025-10-17 16:32:39,994 - INFO - Finished processing chromosome Chr12\n",
      "2025-10-17 16:32:49,753 - INFO - Finished processing chromosome Chr2\n",
      "2025-10-17 16:33:13,071 - INFO - Finished processing chromosome Chr11\n",
      "2025-10-17 16:33:16,015 - INFO - Finished processing chromosome Chr9\n",
      "2025-10-17 16:33:25,115 - INFO - Finished processing chromosome Chr8\n",
      "2025-10-17 16:33:25,597 - INFO - Finished processing chromosome Chr10\n",
      "2025-10-17 16:33:38,271 - INFO - Finished processing chromosome Chr4\n",
      "2025-10-17 16:33:39,502 - INFO - Finished processing chromosome Chr5\n",
      "2025-10-17 16:33:50,289 - INFO - Finished processing chromosome Chr7\n",
      "2025-10-17 16:34:00,427 - INFO - Finished processing chromosome Chr1\n",
      "2025-10-17 16:34:42,777 - INFO - Finished processing chromosome Chr3\n",
      "2025-10-17 16:49:26,430 - INFO - Finished processing chromosome Chr6\n",
      "2025-10-17 16:51:16,803 - INFO - Read assignments are stored in /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/OUT.read_assignments.tsv.gz\n",
      "2025-10-17 16:51:16,803 - INFO - Read assignment statistics\n",
      "2025-10-17 16:51:16,803 - INFO - ambiguous: 378666\n",
      "2025-10-17 16:51:16,803 - INFO - inconsistent: 1249184\n",
      "2025-10-17 16:51:16,803 - INFO - inconsistent_ambiguous: 176031\n",
      "2025-10-17 16:51:16,803 - INFO - inconsistent_non_intronic: 438389\n",
      "2025-10-17 16:51:16,803 - INFO - intergenic: 300565\n",
      "2025-10-17 16:51:16,803 - INFO - noninformative: 1586487\n",
      "2025-10-17 16:51:16,803 - INFO - unique: 1540160\n",
      "2025-10-17 16:51:16,803 - INFO - unique_minor_difference: 135627\n",
      "2025-10-17 16:51:16,803 - INFO - Gene counts are stored in /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/OUT.gene_counts.tsv\n",
      "2025-10-17 16:51:16,804 - INFO - Transcript counts are stored in /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/OUT.transcript_counts.tsv\n",
      "2025-10-17 16:51:16,804 - INFO - Counts can be converted to other formats using src/convert_grouped_counts.py\n",
      "2025-10-17 16:51:16,975 - INFO - Transcript model statistics\n",
      "2025-10-17 16:51:16,975 - INFO - known: 6176\n",
      "2025-10-17 16:51:16,975 - INFO - novel_in_catalog: 756\n",
      "2025-10-17 16:51:16,975 - INFO - novel_not_in_catalog: 9900\n",
      "2025-10-17 16:52:05,298 - INFO - Transcript model file /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/OUT.transcript_models.gtf\n",
      "2025-10-17 16:52:05,448 - INFO - Extended annotation is saved to /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/OUT.extended_annotation.gtf\n",
      "2025-10-17 16:52:05,448 - INFO - Counts for generated transcript models are saves to: /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant/OUT/OUT.discovered_transcript_counts.tsv\n",
      "2025-10-17 16:52:05,822 - INFO - Processed experiment OUT\n",
      "2025-10-17 16:52:05,822 - INFO - Processed 1 experiment\n",
      "2025-10-17 16:52:05,822 - INFO -  === IsoQuant pipeline finished === \n",
      "\u001b[32m2025-10-17 16:52:06.352\u001b[0m | \u001b[32m\u001b[1mSUCCESS \u001b[0m | \u001b[36mspider_silkome_module.features\u001b[0m:\u001b[36mrun_shell_command_with_check\u001b[0m:\u001b[36m53\u001b[0m - \u001b[32m\u001b[1mCommand executed successfully, output file: /home/gyk/project/spider_silkome/data/interim/03.DRS_mapping/isoquant\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation_file = genome_file.replace(\".fa\", \"_fixed_agat.gff\")\n",
    "fastq_file = f\"{RAW_DATA_DIR}/Tclav-F1_Ar-28-Trcl-f/pass.fq.gz\"\n",
    "isoquant_output_dir = f\"{INTERIM_DATA_DIR}/03.DRS_mapping/isoquant\"\n",
    "isoquant_cmd = f\"isoquant.py --genedb {gene_annotation_file} --reference {genome_file} --fastq {fastq_file} --data_type nanopore -o {isoquant_output_dir}\"\n",
    "run_shell_command_with_check(isoquant_cmd, isoquant_output_dir,force=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af55596d",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
