{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eeb5c54c",
   "metadata": {},
   "source": [
    "# Miniprot mapping\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0201bd95",
   "metadata": {},
   "source": [
    "## Configure\n",
    "\n",
    "Setting up the required Python packages\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ebd35e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spider_silkome_module import (\n",
    "    RAW_DATA_DIR,\n",
    "    INTERIM_DATA_DIR,\n",
    "    EXTERNAL_DATA_DIR,\n",
    "    PROCESSED_DATA_DIR\n",
    ")\n",
    "from spider_silkome_module.utils import run_cmd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d9dcd7a",
   "metadata": {},
   "source": [
    "Definition of the path for relevant data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2535d457",
   "metadata": {},
   "outputs": [],
   "source": [
    "task_name = \"miniprot_mapping_20260209\"\n",
    "\n",
    "spidroin_fasta_file = EXTERNAL_DATA_DIR / \"spider-silkome-database.v1.prot.fixed.renamed.fasta\"\n",
    "spider_genome_path = RAW_DATA_DIR / \"01.ref_gff\"\n",
    "\n",
    "# dedup\n",
    "dedup_output_dir = INTERIM_DATA_DIR / task_name / \"cdhit\"\n",
    "spidroin_fasta_file_rep = dedup_output_dir / \"cdhit_rep_seq.fa\"\n",
    "shortest_seq_file = dedup_output_dir / \"cdhit_shortest_seq.fa\"\n",
    "\n",
    "# miniprot\n",
    "gnome_mpi_path = INTERIM_DATA_DIR / task_name / \"gnome_mpi\"\n",
    "miniprot_output_path = PROCESSED_DATA_DIR / task_name / \"miniprot_output\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98575d4b",
   "metadata": {},
   "source": [
    "## cd-hit\n",
    "Perform redundancy reduction on the N/C terminal sequences of Spidroin proteins using cd-hit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96ab84e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cd-hit\n",
    "run_cmd(\n",
    "    f\"pixi run cd-hit -i {spidroin_fasta_file} -o {spidroin_fasta_file_rep} -c 0.9 -T 0 -M 0 -d 0\",\n",
    "    [spidroin_fasta_file_rep]\n",
    ")\n",
    "\n",
    "# extract shortest seq\n",
    "run_cmd(\n",
    "    f\"pixi run python -m spider_silkome_module.extract_shortest_seq \\\n",
    "        --clstr-path {spidroin_fasta_file_rep}.clstr \\\n",
    "        --fasta-path {spidroin_fasta_file} \\\n",
    "        --output-path {shortest_seq_file}\",\n",
    "    [shortest_seq_file]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbb68ca3",
   "metadata": {},
   "source": [
    "## miniprot\n",
    "\n",
    "Use miniprot to align the C-terminal and N-terminal sequences of the spidroin gene sequences to the genome of the new species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b2f321b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genome index\n",
    "run_cmd(\n",
    "    f\"miniprot -t70 -d {gnome_mpi_path}/{spider}.mpi {spider_genome_path}/{spider_genome}\",\n",
    "    [gnome_mpi_path]\n",
    ")\n",
    "\n",
    "\n",
    "# miniprot\n",
    "run_cmd(\n",
    "    f\"pixi run python -m spider_silkome_module.run_miniprot \\\n",
    "        --input-path {spider_genome_path} \\\n",
    "        --protein-fasta {shortest_seq_file} \\\n",
    "        --output-path {miniprot_output_path} \\\n",
    "        --threads 70 --outc 0.8 --force\",\n",
    "    [miniprot_output_path]\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
